### ISSUE TITLE
Configure Mediaflux to honor CAS http headers

### ISSUE THREAD

**Issue by bess** [pulibrary/tiger-data-app#863]
Document to track our attempts at an auth solution for mediaflux: https://docs.google.com/document/d/1jNFKCRjA-vCz01OTNXf5I4rPSmAK7i6pvXbStBtbWCQ/edit?usp=sharing

Sounds like we haven't quite settled on a solution yet. New email from Jason on 31 July, added to feature refinement agenda for 31 July.

> Hi Bess,
>  
> I woke up thinking about the authentication this morning – essentially about security (or lack of it).
>  
> If you pass through the username and we accept that, then we have to trust you have intercepted all HTTP traffic (all) in front of Mediaflux. Are you absolutely confident that you can do that from all IP vectors?
>  
> Another approach – which I want to test is the following:
>  
> You can provide the user name, but also 
> Provide a token that was generated by CAS, that
> We can then use to contact CAS directly to validate the user.
>  
> With the token, we don’t need the username at all really – we could ask CAS for the username from the token. You would them pass the CAS token to us.
>  
> I’m not sure what tokens CAS produces, and whether we could reach out to a (REST? Or other) end-point to get details on the user.
>  
> I feel that would be more secure.
>  
> Is that possible?
>  
> Jason

This library seems relevant: https://github.com/apereo/java-cas-client

My response: 

Hi, Jason.

I really like the approach you’re outlining here. This would basically mean that Mediaflux would be implementing CAS, which is a great solution for us, and I believe would also solve the issue of how to do authentication should we want to publish an API, which has also been requested. 

There is a java-cas library that you could take advantage of: https://github.com/apereo/java-cas-client

If this would work with Princeton’s existing CAS setup, this seems like a great choice. If it’s going to mean a lot of involvement from Princeton’s AIM team, then it might not be a workable option. Our team has proposed an [experiment](https://github.com/pulibrary/tiger-data-app/issues/868) to determine whether we can validate a token in an arbitrary context, which looks like it should be possible, and if we can do that then you should be able to the same thing on the mediaflux side. 

Bess

**Comment by hectorcorrea** [comment]
Direct link to the ticket: https://support.arcitecta.com/hc/en-us/requests/6652 

(you need to have an account with Arcitecta's support system to be able to see it)

**Comment by kelynch** [comment]
Is https://github.com/pulibrary/tiger-data-app/issues/836 a duplicate of this ticket?

**Comment by bess** [comment]
We need to be able to configure mediaflux to honor the http headers that indicate that a user has authenticated to CAS. Arcitecta has promised us an answer, and this is being tracked in their ticketing system under request 6652. Here is the text of that ticket as of 26 July 2024:

> 
> Your request (6652) has been updated. To add additional comments, reply to this email.
> 
> Jason Lohrey (Arcitecta Support)
> Jul 25, 2024, 22:02 GMT+10
> 
> Perfect. I’ll make a final decision and let you know. This will be easy. Thanks for doing that too.
> Jason Lohrey / CTO
> jason.lohrey@arcitecta.com / +61 418 364 941
> 
> Bess O. Sadler
> Jul 25, 2024, 21:55 GMT+10
> 
> Hi, Jason. Thank you so much for taking a look and sending us the needed configuration. We did not know the exact fields you needed, so we took a guess. If you don’t need the time out one, we can certainly remove it. If you want us to combine the other one, we can do that too. Just let us know. Thank you!
> 
> Bess
> 
> Get Outlook for iOS
> Jason Lohrey (Arcitecta Support)
> Jul 24, 2024, 07:31 GMT+10
> 
> Hi Bess,
> 
> I did not realise the ticket was here. That is great - I will, or I will asked someone else to, identify the configuraiton in Mediaflux to recognise those headers. We might (not yet) combine those headers into a single header: e.g. user@domain
> 
> What does Tigerdata_timeout mean? How should we should interpret this?
> 
> In fact, it's probably irrelevant to us.
> 
> If I don't see / respond to a ticket, feel free to poke (email etc.).
> 
> Jason
> 
> Jason Lohrey / CTO
> jason.lohrey@arcitecta.com / +61 418 364 941
> 
> Steven Easter (Arcitecta Support)
> Jul 19, 2024, 06:14 GMT+10
> 
> Adding email from Bess to the ticket: 
>  
> From: Bess O. Sadler <bess.sadler@princeton.edu>
> Sent: Thursday, July 18, 2024 1:46 PM
> To: Robert R. Knight <knight@princeton.edu>; Steven Easter <steven.easter@arcitecta.com>; Steve Paulson <steve.paulson@arcitecta.com>; Graham Beasley <graham.beasley@arcitecta.com>; Kate Lynch <katelynch@Princeton.EDU>; Charles R. Bentler <cbentler@princeton.edu>; Matt Chandler <mchandler@Princeton.EDU>
> Cc: Jason Lohrey <jason.lohrey@arcitecta.com>
> Subject: Next steps for testing CAS authentication via http headers?
> Dear Jason and colleagues,
>  
> We now have a first iteration of adding CAS authentication to http headers sent to mediaflux. You can see my colleague Hector’s actual code changes here if that’s of interest, but I think the relevant bit is in the attached screen shot from wireshark, showing the http headers being sent along with a mediaflux asset.exists query.
>  
> We also now have a local installation of mediaflux, which, as soon as we iron out a few firewall issues, we should be able to configure however you indicate.
>  
> Please, what is the next step? How do we configure mediaflux to recognize and honor these http headers?
>  
> Onward!
>  
> Many thanks,
> Bess
>  
> Attachment(s)
> 2024July18_Wireshark_HTTP_header.png
> 
> Steven Easter (Arcitecta Support)
> Jul 13, 2024, 04:11 GMT+10
> 
> The plan laid out by Jason: 
> 1.      You have a UI that intercepts requests to access your portal
> 
> 2.      That UI manages the authentication with CAS
> 
> 3.      That UI then makes calls to Mediaflux, injecting an agreed variable with the user name into the HTTP headers
> 
>  
> Has further question. 
> Re: your plan for authn/authz. This sounds great to me and I would think we could do this without the involvement of our central IT IAM group (which is good, since we’re having trouble getting their attention). How do we get specific about the http header name(s), and how do we get Mediaflux configured to handle those headers?
> 
>  
> 
> We have this place in the tiger data code where we are building the http request: https://github.com/pulibrary/tiger-data-app/blob/main/app/models/mediaflux/http/request.rb#L152-L170
> 
>  
> 
> I’m thinking we want something like (pseudocode):
> 
>  
> 
>       def build_http_request(name:, form_file: nil)
> 
>               request = self.class.build_post_request
> 
>               request.headers << @current_user.netid
> 
>               request.headers << @current_user.groups
> 
>  
> 
> Because won’t we also need to know what Princeton groups someone is a member of? We can certainly get the netid, and maybe that’s the place to start. We are promised that at some point we will have a system in place for querying group membership, but that’s not in place yet, I just want to check my logic that this is what you’re imagining.
> 
>  
> With the AD User Name, Mediaflux can look up & resolve their group membership. 
>

**Comment by hectorcorrea** [comment]
Once a user has authenticated with CAS we have access to the "ticket":

```
request.env["omniauth.auth"].credentials.ticket
=> "ST-xxxxx-yyyyyyyyyyyyyyyyy-zzzzz-fed"
```

**Comment by kelynch** [comment]
Waiting on information from Arcitecta on new approach (conversation and work with the vendor being tracked here) - https://support.arcitecta.com/hc/en-us/requests/6652

When we get word back, we should then close this ticket and we will open a new ticket and link to it here, then close this ticket.

**Comment by carolyncole** [comment]
close in favor of #881
